File: src/lib/ipc/picus/activity.ts
/**
 * Picus session monitor IPC — what the server is doing, and stopping one of it.
 *
 * A file of its own rather than three more exports in `db.ts`: this is the only
 * part of the surface that describes the *server* rather than a schema or a
 * statement, it is the only one that is polled, and it is the only one that can
 * end somebody else's transaction.
 *
 * The shapes mirror `picus_db_api::activity` field for field — the Rust side
 * serialises camelCase precisely so there is no translation layer here.
 */

import { picus } from '../rpc';

/** One backend connected to the server. */
export interface SessionActivity {
  /** Server-side process id — the handle every action takes. */
  pid: number;
  user: string;
  database: string;
  /** What the client called itself: `psql`, `Picus`, an application name. */
  application: string;
  /** Client address; empty for a connection over the local socket. */
  client: string;
  /** `active`, `idle`, `idle in transaction`, … as the server words it. */
  state: string;
  /** What it is waiting on. `null` means it is running, not that it is idle. */
  waitEvent: string | null;
  /** The statement, as the server holds it — it may have truncated it itself. */
  query: string;
  /**
   * Ages in milliseconds, computed **by the server**. Never derive these from the
   * browser's clock: the two machines disagree, sometimes by hours, and a wrong
   * duration here looks like a finding rather than like a bug.
   */
  queryAgeMs: number | null;
  /** How long it has been in its current state — what identifies an abandoned
   *  `idle in transaction`. */
  stateAgeMs: number | null;
  transactionAgeMs: number | null;
  /** Picus's own session. Legal to stop, but it has to be labelled. */
  isSelf: boolean;
  /** Pids this session is waiting for; empty when it is not blocked. */
  blockedBy: number[];
}

/** One "waiter is stuck behind blocker" relationship. */
export interface BlockEdge {
  waiter: number;
  blocker: number;
  /** The object being contended, when the server names one. */
  relation: string | null;
  /** The lock mode being waited for. */
  mode: string | null;
}

/** Everything the monitor shows, from one read. */
export interface ActivitySnapshot {
  sessions: SessionActivity[];
  /** Who is waiting for whom. Empty when nothing is blocked. */
  blocked: BlockEdge[];
  /** The server's own idea of when this was read, as it formats it. Displayed,
   *  never computed with. */
  readAt: string;
}

/**
 * How firmly to ask a session to stop.
 *
 * Two verbs, not one: `cancel` asks the *statement* to stop and leaves the
 * connection alive, which is almost always what is wanted; `terminate` drops the
 * connection and **rolls its transaction back**, and is the answer only for a
 * session that is not running anything — the abandoned `idle in transaction`
 * holding a lock.
 */
export type StopKind = 'cancel' | 'terminate';

/** One coherent picture of the server: sessions plus the blocking graph. */
export function readActivity(connectionId: string): Promise<ActivitySnapshot> {
  return picus('picus_activity', { id: connectionId });
}

/**
 * Ask one backend to stop.
 *
 * Resolves `false` when the server found no such pid — the session ended between
 * the read and the click, which is ordinary and not an error. A refusal for want
 * of privilege **rejects**, carrying the server's own sentence: "you may not" and
 * "it was already gone" must not look the same to whoever pressed the button.
 */
export function stopSession(
  connectionId: string,
  pid: number,
  kind: StopKind,
): Promise<boolean> {
  return picus('picus_stop_session', { id: connectionId, pid, kind });
}

File: crates/products/picus/db-provider/api/src/activity.rs
//! What a server is doing **right now** — the session monitor's vocabulary.
//!
//! This is the one part of the API that describes the server rather than the
//! schema, and it exists because of a failure that actually happened: a statement
//! stopped responding, Cancel did nothing, and from inside Picus there was no way
//! to see whether the backend was still running, waiting on a lock, or already
//! gone. A tool that opens sessions on other people's databases has to be able to
//! answer "what is stuck, and what is holding it".
//!
//! Two deliberate shapes here:
//!
//! * **Blocking is an edge list, not a flag.** "This session is blocked" is not
//!   actionable; "this session is blocked by pid 4412, which is idle in a
//!   transaction" is. The graph is what lets the interface draw the chain to the
//!   session at the root of it, which is the only one worth acting on.
//! * **Times are ages in milliseconds, not timestamps.** The server's clock and
//!   the client's disagree, sometimes by hours, and a duration computed by
//!   subtracting one from the other is wrong in a way nobody notices. The server
//!   computes the age; we carry it.

use serde::{Deserialize, Serialize};

/// Everything the monitor shows in one read.
///
/// One read, not two: the sessions and the blocking edges have to describe the
/// same instant or the interface draws an arrow from a session to one that has
/// already finished.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivitySnapshot {
    pub sessions: Vec<SessionActivity>,
    /// Who is waiting for whom. Empty when nothing is blocked.
    pub blocked: Vec<BlockEdge>,
    /// The server's own idea of when this was read, as it formats it. Displayed,
    /// never computed with.
    pub read_at: String,
}

/// One backend connected to the server.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionActivity {
    /// Server-side process id. The handle every action takes.
    pub pid: i32,
    pub user: String,
    pub database: String,
    /// What the client called itself — `psql`, `Picus`, an application name.
    pub application: String,
    /// Client address, or empty for a local socket.
    pub client: String,
    /// `active`, `idle`, `idle in transaction`, … as the server words it.
    pub state: String,
    /// What it is waiting on, when it is waiting. `None` means it is running.
    pub wait_event: Option<String>,
    /// The statement, as the server holds it. May be truncated by the server's own
    /// `track_activity_query_size`; that is the server's truncation, not ours.
    pub query: String,
    /// How long the current statement has been running, in milliseconds.
    pub query_age_ms: Option<i64>,
    /// How long the session has been in its current state, in milliseconds. This is
    /// the number that identifies an abandoned `idle in transaction`.
    pub state_age_ms: Option<i64>,
    /// How long the transaction has been open, in milliseconds.
    pub transaction_age_ms: Option<i64>,
    /// This is Picus's own session. Killing it is legal and occasionally what the
    /// user wants, but it has to be labelled rather than discovered.
    pub is_self: bool,
    /// Pids this session is waiting for. Denormalised from [`ActivitySnapshot::blocked`]
    /// so a row can be rendered without a lookup.
    pub blocked_by: Vec<i32>,
}

/// One "waiter is stuck behind blocker" relationship.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BlockEdge {
    pub waiter: i32,
    pub blocker: i32,
    /// The object being contended, when the server names one.
    pub relation: Option<String>,
    /// The lock mode being waited for.
    pub mode: Option<String>,
}

/// How firmly to ask a session to stop.
///
/// Two verbs and not one because they are genuinely different acts, and merging
/// them would make the gentler one unreachable: cancelling asks the *statement* to
/// stop and leaves the connection alive, which is almost always what is wanted;
/// terminating drops the connection, rolls its transaction back, and is the answer
/// only for a session that is not running anything at all — the classic abandoned
/// `idle in transaction` holding a lock.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum StopKind {
    /// Cancel the running statement; the session survives.
    Cancel,
    /// Close the connection outright.
    Terminate,
}
